iot-hub/Samples/service/CleanupDevicesSample/CleanupDevicesSample.cs

```diff
@@ -0,0 +1,168 @@
+                foreach (string prefix in _deleteDeviceWithPrefix)
+                {
+                    if (deviceId.StartsWith(prefix, StringComparison.OrdinalIgnoreCase))
+                    {
+                        devicesToBeDeleted.Add(device);
+                    }
+                }
+            }
+            devicesToBeDeleted
+                .ForEach(device => device.ImportMode = ImportMode.Delete);
+            Console.WriteLine($"Retrieved {devicesToBeDeleted.Count} devices for deletion.");
+
+            if (devicesToBeDeleted.Count > 0)
+            {
+                // Step 3a: Write the new import data back to the blob.
+                Stream devicesFile = ImportExportDevicesHelpers.BuildDevicesStream(devicesToBeDeleted);
+                await blobClient.UploadAsync(devicesFile, overwrite: true);
+
+                // Step 3b: Call import using the same blob to delete all devices.
+                JobProperties importDevicesToBeDeletedProperties = JobProperties
+                    .CreateForImportJob(
+                        inputBlobContainerUri: storageAccountSasUri,
+                        outputBlobContainerUri: storageAccountSasUri,
+                        storageAuthenticationType: StorageAuthenticationType.KeyBased);
+                JobProperties importDevicesToBeDeletedJob = await _registryManager.ImportDevicesAsync(importDevicesToBeDeletedProperties);
+
+                // Wait until job is finished.
+                while (true)
+                {
+                    importDevicesToBeDeletedJob = await _registryManager.GetJobAsync(importDevicesToBeDeletedJob.JobId);
+                    if (s_completedJobs.Contains(importDevicesToBeDeletedJob.Status))
+                    {
+                        // Job has finished executing.
+                        break;
+                    }
+
+                    Console.WriteLine($"Job {importDevicesToBeDeletedJob.JobId} is {importDevicesToBeDeletedJob.Status} with progress {importDevicesToBeDeletedJob.Progress}%");
+                    await Task.Delay(_waitDuration);
+                }
+                Console.WriteLine($"Job {importDevicesToBeDeletedJob.JobId} is {importDevicesToBeDeletedJob.Status}.");
+            }
+
+            // Step 4: Delete the storage container created.
+            await _blobContainerClient.DeleteAsync();
+            Console.WriteLine($"Storage container {_blobContainerClient.Name} deleted.");
+        }
+
+        private async Task PrintDeviceCountAsync()
+        {
+            string countSqlQuery = "SELECT COUNT() AS numberOfDevices FROM devices";
+            IQuery countQuery = _registryManager.CreateQuery(countSqlQuery);
+            while (countQuery.HasMoreResults)
+            {
+                IEnumerable<string> result = await countQuery.GetNextAsJsonAsync();
+                foreach (var item in result)
+                {
+                    Console.WriteLine($"Total # of devices in the hub: \n{item}");
```

Comment:
Is this foreach required? Just need to keep count.

Reply:
the result is an IEnumerable, so we will need the for-each. For this specific query we know that result.Count = 1, but I didn't want to deviate from query-result patterns. That said, I don't have a strong preference for wanting to have it this way.

Reply:
I think you know it is going to be a single result, because you used an aggregate in the query string, so you can just grab the first item.

Reply:
Interesting, that was confusing at first. So the result count is always 1 but return type is IEnumerable<string>!!

Also it seems like we are printing count of all devices and not only the ones with _deleteDeviceWithPrefix, correct me if that's wrong.

Reply:
Ok so it not as part of this method PrintDeviceCountAsync, why do we then need this method? OR is is there to demonstrate how to use query API?

Reply:
> Ok so it not as part of this method PrintDeviceCountAsync, why do we then need this method?

There are 2 counts in question here - the total no of devices on your hub and the no of devices being deleted. The reason we print both of those is for us to compare and verify that we are indeed deleting all "expected" devices that fit the criteria (device prefixes that are supplied as an input to this application). If we print only the count of devices being deleted, we wouldn't know if our input file/ deletion logic is incorrect unless we look at the hub instance and verify the device registry.

> OR is is there to demonstrate how to use query API?

This is not the intention, no.

Reply:
That would make sense if there is any logic of comparing these counts to validate. Shouldn't logging devices id which are being deleted also be sufficient since those are the ones with prefix? But either ways, I don't have a strong preference for the reasoning you mentioned so I'll leave it to you for decision.

If kept as it is, it would be ideal to rename it to PrintTotalDevicesCountAsync or add a comment for clarity with reasoning to why you think it needs to be printed. As you mentioned earlier .... to verify that we are indeed deleting all "expected" devices that fit the criteria (device prefixes that are supplied as an input to this application)